File: src/main/java/io/github/eone666/telegramnotifier/TelegramNotifier.java
package io.github.eone666.telegramnotifier;

import io.github.eone666.telegramnotifier.events.PlayerJoin;
import io.github.eone666.telegramnotifier.events.PlayerQuit;
import io.github.eone666.telegramnotifier.features.Notifications;
import io.github.eone666.telegramnotifier.features.PlayersList;
import io.github.eone666.telegramnotifier.utils.Config;
import io.github.eone666.telegramnotifier.utils.telegram.Telegram;
import org.bukkit.plugin.java.JavaPlugin;

public final class TelegramNotifier extends JavaPlugin {

    public Config config = new Config(this);
    public Telegram tg = new Telegram(config.getToken());
    public Notifications notifications = new Notifications(this);
    public PlayersList playersList = new PlayersList(this);

    @Override
    public void onEnable() {
        saveDefaultConfig();
        //init features
        playersList.init();
        //register events
        getServer().getPluginManager().registerEvents(new PlayerJoin(this), this);
        getServer().getPluginManager().registerEvents(new PlayerQuit(this), this);

        getLogger().info("Started successfully");
    }

    @Override
    public void onDisable() {
         getLogger().info("Shutting down...");
         //disable features
         playersList.disable();
    }
}

File: src/main/kotlin/io/github/eone666/telegramnotifier/TelegramNotifier.kt
package io.github.eone666.telegramnotifier

import org.bukkit.plugin.java.JavaPlugin
import com.github.shynixn.mccoroutine.bukkit.registerSuspendingEvents


import io.github.eone666.telegramnotifier.listeners.PlayerJoin
import io.github.eone666.telegramnotifier.listeners.PlayerQuit

import io.github.eone666.telegramnotifier.features.Notifications
import io.github.eone666.telegramnotifier.features.PlayerList
import io.github.eone666.telegramnotifier.features.OneTimePasswordForSender

import io.github.eone666.telegramnotifier.telegram.Bot

import io.github.eone666.telegramnotifier.commands.Setup as SetupCommand
import io.github.eone666.telegramnotifier.commands.CancelSetup as CancelSetupCommand
import io.github.eone666.telegramnotifier.commands.Config as ConfigCommand

class TelegramNotifier : JavaPlugin() {
    lateinit var config: Config
    var bot: Bot? = null
    lateinit var oneTimePasswordForSender: OneTimePasswordForSender
    lateinit var notifications: Notifications
    lateinit var playerList: PlayerList

    fun initFeatures() {
        notifications = Notifications()
        playerList = PlayerList()

        //init features
        playerList.update()
    }

    override fun onEnable() {
        instance = this

        config = Config()

        if (config.isPluginConfigured.get()) {
            logger.info("Plugin configured, starting bot...")
            bot = Bot(config.token.get())
            initFeatures()
        }

        //register events
        server.pluginManager.registerSuspendingEvents(PlayerJoin(), this)
        server.pluginManager.registerSuspendingEvents(PlayerQuit(), this)

        //register minecraft commands
        getCommand("setup")?.setExecutor(SetupCommand())
        getCommand("cancelsetup")?.setExecutor(CancelSetupCommand())
        getCommand("config")?.setExecutor(ConfigCommand())

        logger.info("Started successfully")
    }

    override fun onDisable() {
        logger.info("Shutting down...")
        bot?.stop()

        // disable features
        playerList.clear()
    }

    companion object {
        lateinit var instance: TelegramNotifier
    }
}

val pluginInstance: TelegramNotifier by lazy { TelegramNotifier.instance }